enhancements/microshift/generic-device-plugin-for-microshift.md

```diff
@@ -0,0 +1,461 @@
+	Type PathType `json:"type"`
+	// Limit specifies up to how many times this device can be used in the group concurrently when other devices
+	// in the group yield more matches.
+	// For example, if one path in the group matches 5 devices and another matches 1 device but has a limit of 10,
+	// then the group will provide 5 pairs of devices.
+	// When unspecified, Limit defaults to 1.
+	Limit uint `json:"limit,omitempty"`
+}
+```
+
+**PathType**
+```go
+// PathType represents the kinds of file-system nodes that can be scheduled.
+type PathType string
+
+const (
+	// DevicePathType represents a file-system device node and is mounted as a device.
+	DevicePathType PathType = "Device"
+	// MountPathType represents an ordinary file-system node and is bind-mounted.
+	MountPathType PathType = "Mount"
+)
+```
+
+**USBSpec**
+```go
+// USBSpec represents a USB device specification that should be discovered.
+// A USB device must match exactly on all the given attributes to pass.
+type USBSpec struct {
+	// Vendor is the USB Vendor ID of the device to match on.
+	// (Both of these get mangled to uint16 for processing - but you should use the hexadecimal representation.)
+	Vendor USBID `json:"vendor"`
+	// Product is the USB Product ID of the device to match on.
+	Product USBID `json:"product"`
+	// Serial is the serial number of the device to match on.
+	Serial string `json:"serial"`
+}
+```
+
+### Topology Considerations
+
+#### Hypershift / Hosted Control Planes
+
+N/A
+
+#### Standalone Clusters
+
+N/A
+
+#### Single-node Deployments or MicroShift
+
+Enhancement is intended for MicroShift and it will extend MicroShift configuration.
+
+### Implementation Details/Notes/Constraints
+
+It was decided that initially the Generic Device Plugin will not be forked.
+Instead we'll work with the upstream on contributing changes that would be useful to us.
+Potentially but not necessarily:
+- Switch to the `klog` to avoid importing another logging library by the MicroShift.
+- Restructure `main` module into submodules for easier reuse within MicroShift
+  (instead of copying the contents of `main`).
+
+"Code import" approach was chosen to avoid the effort associated with
+productizing a new image. The MicroShift team at this time does not own any
+images and the RPMs are managed by ART. See "Alternatives" section for "container" approach.
+
+- Importing the code as controller into the MicroShift binary means more or less
+  replicating the functionality of the `main()` function.
+- A duplicate of GDP's configuration structs may be used in MicroShift's 
+  configuration code to make it more readable if needed.
+- GDP will be an optional feature, disabled by default, but included within
+  the MicroShift binary.
+- MicroShift will expose the devices with `device.microshift.io/NAME` label
+  (changed via configuration from `squat.ai/NAME`).
+
+### Risks and Mitigations
+
+Not forking means that time sensitive fixes (like security) will need to be vendor-patched.
```

Comment:
Meaning on a device-specific basis and outside our release process?

Reply:
Ah no, it comes down to the fact that we're using something called "vendoring" which is basically mechanism that Go dependencies are copied into a vendor/ dir in our repository.
So if we need, we can patch the code we imported from upstream to the https://github.com/openshift/microshift/tree/main/vendor